### ISSUE TITLE
Hello, Can I use HPHI to calculate the dynamic spin structure factor? What should I do?

### ISSUE THREAD

**Issue by willOcean1** [issp-center-dev/HPhi#184]
Can you give a simple example to calculate the dynamic spin structure factor

**Comment by ultimatile** [comment]
A tutorial for the dynamical spin structure factor is provided at https://issp-center-dev.github.io/HPhi/manual/develop/tutorial/en/html/dynamical/spectrum.html.
As the document says, the sample input files are available as `samples/tutorial_4.1/stan1.in` and `samples/tutorial_4.1/stan2.in` in the `samples` directory of HPhi's source files.

**Comment by willOcean1** [comment]
I want to calculate the dynamic spin structure factor  using expert mode. Is this possible? I need to calculate the DSSF for different values of q and w. What should I do?

**Comment by willOcean1** [comment]
Can I directly calculate 
𝑆(𝑘,𝑤)=𝑆𝑥(𝑘,𝑤)+𝑆𝑦(𝑘,𝑤)+𝑆𝑧(𝑘,𝑤) 
using the expert mode, if I generate all possible cases in the excitation file? For example, the content of the excitation file for a 2-site system is as shown in the diagram. Or do I only calculate one component of S(k,w) at a time?
![image](https://github.com/user-attachments/assets/0f53965b-03a7-4fe8-9e68-1686f3a4a911)

**Comment by ultimatile** [comment]
As far as I know, it’s impossible to sum them up within HPhi. Please calculate your desired observables yourself from the raw data generated by HPhi.

**Comment by willOcean1** [comment]
"Hello, I made some changes to the Python script in tutorial 4.3, but the graph I plotted has some issues. Some of the calculated DSSF values are 0, and these values of q are not fixed; they change every time I run the program. What could be the reason for this?"
![pic](https://github.com/user-attachments/assets/c75622db-46ee-40d9-9479-60434b5f2c63)
[aaaa.txt](https://github.com/user-attachments/files/17459308/aaaa.txt)

**Comment by willOcean1** [comment]
"Hello, I made some changes to the Python script in tutorial 4.3, but the graph I plotted has some issues. Some of the calculated DSSF values are 0, and these values of q are not fixed; the "q" change every time I run the program. What could be the reason for this?"  
![pic](https://github.com/user-attachments/assets/021d50d7-675d-40ea-8ee3-2a5c22dfc9bd)
[spinchain_example.txt](https://github.com/user-attachments/files/17459383/spinchain_example.txt)

**Comment by k-ido** [comment]
@willOcean1 
Could you tell me the version of HPhi?
I tried to reproduce your result using your attached code, but the result is different from yours and seems to be correct.
To obtain the following result, I use v.3.5.2 of HPhi and Python 3.9.7 on Apple M1 MacBook.

![image](https://github.com/user-attachments/assets/b2229818-32f9-4c95-8e38-56bb1c009be8)

**Comment by willOcean1** [comment]
I ran the Python script again on my computer, but the results are still incorrect. I've noticed that some "spectrumi.dat" files contain second column numbers that approach negative infinity. I'm using HPhi version 3.5.2 and Python 3.8.10. However, when I ran this script on a supercomputing platform, the results were correct, using HPhi version 3.5.2 and Python 3.6.8.
/home/lius/Documents/Re/Re_DSSF/src/CCC/spectrum1.txt

**Comment by willOcean1** [comment]
I ran the Python script again on my computer, but the results are still incorrect. I've noticed that some "spectrumi.dat" files contain second column numbers that approach negative infinity. I'm using HPhi version 3.5.2 and Python 3.8.10. However, when I ran this script on a supercomputing platform, the results were correct, using HPhi version 3.5.2 and Python 3.6.8.
[spectrum1.txt](https://github.com/user-attachments/files/17486257/spectrum1.txt)

**Comment by tmisawa** [comment]
I have also tried to identify the origin of the error. However, in my environment (M1Mac, HPhi ver 3.5.2, Python  3.12.5), the script seems to work well.

**Comment by tmisawa** [comment]
@willOcean1 
How did you run the script?  According to the tutorial, just execute `python spinchain_example.py`.
Then, you will obtain `spectrum.dat` with three columns. 
The format of the attached file `spectrum1.txt` seems to be different from that of `spectrum.dat`. 
How do you obtain `spectrum1.txt`?

**Comment by tmisawa** [comment]
@willOcean1 
Sorry, I misunderstood that `specrum1.txt` is `spectrum.dat`. It turns out that `specrum1.txt` is actually a copy of `zvo_DynamicalGreen.dat`. 
Based on the `specrum1.txt`, it seems there might be some problems with the values of the frequency ω. 

In the standard output, there is information on the frequency as follows: 
>Frequency range:
  Omega Max. :     0.00000e+00     1.00000e-01
  Omega Min. :    -1.00000e+01     1.00000e-01
  Num. of Omega : 100

Could you check whether there is a problem with the frequency?

**Comment by willOcean1** [comment]
The frequency is correct.
[modpara.txt](https://github.com/user-attachments/files/17487995/modpara.txt)

**Comment by tmisawa** [comment]
No, please check the information in the **standard output**, not the input file. There may be a possibility that the parameters are not being initialized properly within the program, and I would like to confirm that.

**Comment by willOcean1** [comment]
I'm not sure if I understood you correctly. When running the Python script this time, there were errors with “spectrum3.dat” and “spectrum7.dat”, but the standard output seems to be correct.
![image](https://github.com/user-attachments/assets/7a4543e7-4883-4883-8555-6a5a63d7b6be)

**Comment by tmisawa** [comment]
Thank you for the information about omega. The imaginary parts of omega_max and omega_min
seem to overflow (e+303 is an unusually large number).

There may be a problem with the definition of omega in the program.
To identify the origin of the problem, could you provide the information on the OS and compiler you used?

**Comment by tmisawa** [comment]
We now become aware that there is a bug during file reading in HPhi, i.e., reading` OmegaMax` and `OmegaMin` in `modepara.def`.  Even when the imaginary parts of them are not defined in `modepara.def`, HPhi still attempts to read the imaginary parts of them.
It seems that this does not cause serious problems in most cases, 
but something unusual occurs in your environment. 
We apologize for the inconvenience.

For now, to avoid this problem, please try adding an imaginary part of 0.0 to both `OmegaMax` and `OmegaMin`.

**Comment by willOcean1** [comment]
Yes, it was caused by this issue
